File: apps/api-rs/examples/inspect_document.rs
use std::path::Path;

use anyhow::{Context, Result};
use documind::document;
use serde::Serialize;
use uuid::Uuid;

#[derive(Serialize)]
struct DocumentInspection {
    file_name: String,
    file_type: String,
    pages: Option<i32>,
    blocks: usize,
    tables: usize,
    chunks: usize,
    anchors: usize,
    unanchored_blocks: usize,
    bbox_anchors: usize,
    quality_score: f64,
    warnings: Vec<String>,
    content: String,
}

fn main() -> Result<()> {
    let paths = std::env::args().skip(1).collect::<Vec<_>>();
    if paths.is_empty() {
        anyhow::bail!("usage: cargo run -p documind --example inspect_document -- <files...>");
    }

    for path in paths {
        let path = Path::new(&path);
        let file_name = path
            .file_name()
            .and_then(|name| name.to_str())
            .context("fixture file name is not UTF-8")?;
        let bytes = std::fs::read(path)
            .with_context(|| format!("failed to read fixture {}", path.display()))?;
        let bundle = document::parse_document(
            Uuid::new_v4(),
            Uuid::new_v4(),
            file_name,
            mime_type(path),
            &bytes,
        )
        .with_context(|| format!("failed to parse fixture {}", path.display()))?;
        let inspection = DocumentInspection {
            file_name: file_name.to_string(),
            file_type: bundle.file_type.as_str().to_string(),
            pages: bundle.parsed.pages,
            blocks: bundle.parsed.blocks.len(),
            tables: bundle.parsed.tables.len(),
            chunks: bundle.chunks.len(),
            anchors: bundle.parsed.anchors.len(),
            unanchored_blocks: bundle
                .parsed
                .blocks
                .iter()
                .filter(|block| block.anchor_ids.is_empty())
                .count(),
            bbox_anchors: bundle
                .parsed
                .anchors
                .iter()
                .filter(|anchor| anchor.bbox.is_some())
                .count(),
            quality_score: bundle.parsed.quality_score,
            warnings: bundle.parsed.warnings,
            content: bundle
                .cleaned_blocks
                .iter()
                .filter(|block| !block.is_removed)
                .map(|block| block.cleaned_text.as_str())
                .collect::<Vec<_>>()
                .join("\n"),
        };
        println!("{}", serde_json::to_string(&inspection)?);
    }
    Ok(())
}

fn mime_type(path: &Path) -> &'static str {
    match path.extension().and_then(|ext| ext.to_str()) {
        Some("pdf") => "application/pdf",
        Some("docx") => "application/vnd.openxmlformats-officedocument.wordprocessingml.document",
        Some("pptx") => "application/vnd.openxmlformats-officedocument.presentationml.presentation",
        Some("md" | "markdown") => "text/markdown",
        Some("txt") => "text/plain",
        _ => "application/octet-stream",
    }
}

File: apps/api-bun/scripts/inspect-document.ts
// 移植自 apps/api-rs/examples/inspect_document.rs —— 解析语料巡检（每文件输出一行 JSON）
// 用法: bun run apps/api-bun/scripts/inspect-document.ts <files...>
import { readFileSync } from 'node:fs';
import { basename, extname } from 'node:path';
import { newUuid } from '../src/infra/uuid.ts';
import { parseDocument } from '../src/document/mod.ts';

interface DocumentInspection {
  file_name: string;
  file_type: string;
  pages: number | null;
  blocks: number;
  tables: number;
  chunks: number;
  anchors: number;
  unanchored_blocks: number;
  bbox_anchors: number;
  block_types: Record<string, number>;
  quality_score: number;
  warnings: string[];
  content: string;
}

function mimeType(path: string): string {
  switch (extname(path).slice(1).toLowerCase()) {
    case 'pdf': return 'application/pdf';
    case 'docx': return 'application/vnd.openxmlformats-officedocument.wordprocessingml.document';
    case 'pptx': return 'application/vnd.openxmlformats-officedocument.presentationml.presentation';
    case 'md': case 'markdown': return 'text/markdown';
    case 'txt': return 'text/plain';
    default: return 'application/octet-stream';
  }
}

const paths = process.argv.slice(2);
if (paths.length === 0) {
  console.error('usage: bun run apps/api-bun/scripts/inspect-document.ts <files...>');
  process.exit(1);
}

for (const path of paths) {
  const fileName = basename(path);
  const bytes = new Uint8Array(readFileSync(path));
  const bundle = await parseDocument(
    newUuid(), newUuid(), fileName, mimeType(path), bytes,
  );
  const inspection: DocumentInspection = {
    file_name: fileName,
    file_type: bundle.file_type,
    pages: bundle.parsed.pages,
    blocks: bundle.parsed.blocks.length,
    tables: bundle.parsed.tables.length,
    chunks: bundle.chunks.length,
    anchors: bundle.parsed.anchors.length,
    unanchored_blocks: bundle.parsed.blocks.filter((block) => block.anchor_ids.length === 0).length,
    bbox_anchors: bundle.parsed.anchors.filter((anchor) => anchor.bbox != null).length,
    block_types: bundle.parsed.blocks.reduce<Record<string, number>>((counts, block) => {
      counts[block.block_type] = (counts[block.block_type] ?? 0) + 1;
      return counts;
    }, {}),
    quality_score: bundle.parsed.quality_score,
    warnings: bundle.parsed.warnings,
    content: bundle.cleaned_blocks
      .filter((block) => !block.is_removed)
      .map((block) => block.cleaned_text)
      .join('\n'),
  };
  console.log(JSON.stringify(inspection));
}
